### PR TITLE
First-time tour

### PR DESCRIPTION
**Pull request by FidelisClayton**
# First-time tour

Story view tour

### CODE REVIEW

File: app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -26,6 +26,15 @@ def disable_two_factor
     end
   end
 
+  def tour
```
Comment (FidelisClayton) [inline app/controllers/registrations_controller.rb]: I tried to put this action on `users_controller` but when i try to perform a `PUT` request this redirects the request to `/teams`. How can i fix this and move this action to `users_controller`?

File: app/models/user.rb [hunk 2]
```diff
@@ -7,7 +9,69 @@ class User < ActiveRecord::Base
   gravtastic default: 'identicon'
 
   # FIXME - DRY up, repeated in Story model
-  JSON_ATTRIBUTES = ["id", "name", "initials", "username", "email"]
+  JSON_ATTRIBUTES = ["id", "name", "initials", "username", "email", "tour", "tour_steps"]
```
Comment (FidelisClayton) [inline app/models/user.rb]: @eddiecode, @talyssonoc, @daniloisr Is this the best place to put it?

File: app/models/user.rb [hunk 3]
```diff
@@ -7,7 +9,69 @@ class User < ActiveRecord::Base
...
+  TOUR_STEPS = [
```
Comment (FidelisClayton) [inline app/models/user.rb]: @eddiecode, @talyssonoc, @daniloisr Is this the best place to put it?

File: app/assets/javascripts/tour.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
+const Tether = require('tether');
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: We're not using using Babel on the project yet (@eddiecode  is including it in a PR he's working) so we should avoid const and lambda until his PR is merged.

File: app/assets/javascripts/tour.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  start: function() {
+    let tour = new Shepherd.Tour({
+      defaults: {
+        classes: 'shepherd-theme-arrows',
+        scrollTo: false
+      }
+    });
+
+    function generateButtons(index, user) {
+      let buttons = [];
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: Regarding what I said about const and lambda, also we should avoid let by now.

File: app/models/user.rb (hunk 3 above)
Comment (talyssonoc) [inline app/models/user.rb]: No, we should not mix a domain class (the User) with the tour. You could have a class called `WelcomeTour` inside models and have it there, not all models must be extensions of ActiveRecord 😄

File: app/models/user.rb [hunk 6]
```diff
@@ -7,7 +9,69 @@ class User < ActiveRecord::Base
...
+      attachTo: '.step-chilly-bin bottom',
+      done: false
+    },
+    {
+      title: 'Backlog',
+      text: 'The backlog MUST have all stories by order of PRIORITY. Most valuable and ready-to-start stories at the top and optional or uncertain stories at the very bottom',
+      attachTo: '.step-backlog bottom',
+      done: false
+    },
+    {
```
Comment (talyssonoc) [inline app/models/user.rb]: This should be on the locales.

File: app/models/user.rb [hunk 7]
```diff
@@ -47,6 +111,10 @@ def set_reset_password_token
     raw
   end
 
+  def set_tour_steps
+    self.tour_steps = TOUR_STEPS.to_json
```
Comment (talyssonoc) [inline app/models/user.rb]: This shouldn't be here either.

Comment (FidelisClayton) [inline app/models/user.rb]: What should I do with it?

Comment (talyssonoc) [inline app/models/user.rb]: It could be in the WelcomeTour model.

File: app/assets/javascripts/central.js [hunk 8]
```diff
@@ -61,5 +62,9 @@ var Central = module.exports = {
 
       window.projectView = view;
     });
+
+    setTimeout(function() {
```
Comment (talyssonoc) [inline app/assets/javascripts/central.js]: Since you don't use `this` inside `firstTimeTour.start`, you could do just: `setTimeout(firstTimeTour.start, 200)` (notice we're not _calling_ start, we're passing the function itself to `setTimeout`)

File: app/assets/javascripts/templates/column.ejs [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<div class="column_header">
+<div class="column_header step-<%= id.replace("_", "-") %>">
```
Comment (talyssonoc) [inline app/assets/javascripts/templates/column.ejs]: Why not use a data attribute `data-step="<%= id.replace("_", "-") %>"`?, it seems much cleaner, HTML classes should not be used to holding data, they're for styling 😄 

Also, it would be good to have a method on the column model to remove this `replace` from the template, something like `htmlFriendlyId()`

File: app/assets/javascripts/templates/project_velocity.ejs [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<span class="velocity-value" id="velocity_value">
-  <i class="mi md-18">trending_up</i>
+<span id="velocity_value" class="step-velocity">
+  <i class="mi mi-18">trending_up</i>
```
Comment (talyssonoc) [inline app/assets/javascripts/templates/project_velocity.ejs]: Why the change from `md` to `mi`?

File: app/assets/javascripts/tour.js [hunk 11]
```diff
@@ -0,0 +1,92 @@
+var Tether = require('tether');
+var Shepherd = require('tether-shepherd');
+var axios = require('axios');
+
+function getCurrentUser() {
+  return axios.get('/users/current');
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: Are we sure the client has support to promises here?

File: app/assets/javascripts/tour.js [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    user.tour = false;
+
+  var data = {
+    user: {
+      tour: user.tour,
+      tour_steps: JSON.stringify(user.tour_steps)
+    }
+  };
+
+  return axios.put(`/users/${user.id}/tour`, data);
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: Those requests should be on a separate file, maybe on a `Tour` Backbone model to keep it organized, what do you think?

File: app/assets/javascripts/tour.js (hunk 11 above)
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: Also, I think the user model could be used here.

File: app/assets/javascripts/tour.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+          updateTour(user);
+          tour.next();
+        }
+      });
+
+      return buttons;
+    }
+
+    function setupTour(user, tour) {
+      user.tour_steps.map((step, index) => {
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: We're not using the `map` result here, so it should be a `forEach` to avoid creating a new array, which would make the performance decrease.

File: app/assets/javascripts/tour.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        var tourStep = Object.assign({}, step);
+
+        var selector = step.attachTo.split(" ")[0];
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: This could be in a separate function, it's not clear why the split is needed here.

File: app/assets/javascripts/tour.js [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+          tourStep.buttons = generateButtons(index, user);
+
+          tour.addStep(tourStep.title.toLowerCase().replace(' ', '_'), tourStep);
+        }
+      });
+    };
+
+    function getCurrentUserSuccess(res) {
+      var user = res.data.user;
+      user.tour_steps = JSON.parse(user.tour_steps);
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: Why is that needed? Doesn't it come from the API? It should be parsed already.

File: app/controllers/registrations_controller.rb [hunk 16]
```diff
@@ -26,6 +26,26 @@ def disable_two_factor
...
+    @user = User.find(params[:id])
+    if @user.update(params.require(:user).permit(:tour, :tour_steps))
```
Comment (talyssonoc) [inline app/controllers/registrations_controller.rb]: The `params.require...permit...` call could be in a separate method

File: app/controllers/registrations_controller.rb [hunk 17]
```diff
@@ -26,6 +26,26 @@ def disable_two_factor
...
+      render json: @user, status: :ok
+    else
+      render json: @user.errors, status: :unprocessable_entity
+    end
+  end
+
+  def reset_tour
+    @user = User.find(params[:id])
+    if @user.update(tour: true, tour_steps: WelcomeTour::STEPS.to_json)
+      flash[:notice] = t('reset_tour_success')
```
Comment (talyssonoc) [inline app/controllers/registrations_controller.rb]: You can make it better with just `redirect_to :back, flash: { notice: t('reset_tour_success') }`

File: app/views/layouts/application.html.erb [hunk 18]
```diff
@@ -115,5 +115,10 @@
   <%= javascript_include_tag :attachinary %>
   <%= yield :charts_data %>
   <%= cloudinary_js_config %>
+
+  <script type="text/javascript">
```
Comment (talyssonoc) [inline app/views/layouts/application.html.erb]: Why is this needed?

File: app/views/projects/index.html.erb [hunk 19]
```diff
@@ -1,6 +1,6 @@
 <% content_for :title_bar do %>
   <% if policy(Project).new? %>
-    <%= link_to t('projects.new project'), new_project_path, class: 'btn btn-primary btn-sm navbar-btn' %>
+    <%= link_to t('projects.new project'), new_project_path, class: 'btn btn-primary btn-sm navbar-btn step-create-project' %>
```
Comment (talyssonoc) [inline app/views/projects/index.html.erb]: Same thing here about using `data-step`

File: app/views/projects/index.html.erb [hunk 20]
```diff
@@ -9,7 +9,7 @@
     <div class="col-md-12">
       <% unless @projects.empty? %>
 
-        <div class="col-md-6">
+        <div class="col-md-6 activity step-activity">
```
Comment (talyssonoc) [inline app/views/projects/index.html.erb]: Same thing here about using `data-step`

File: app/views/registrations/edit.html.erb [hunk 21]
```diff
@@ -33,6 +33,9 @@
         </div>
         <hr />
         <div class="account-options">
+          <% if @user.tour == false %>
```
Comment (talyssonoc) [inline app/views/registrations/edit.html.erb]: You can use just `unless @user.tour`

File: app/views/registrations/edit.html.erb [hunk 22]
```diff
@@ -33,6 +33,9 @@
...
+            <%= link_to t('reset_tour'), "/users/#{@user.id}/reset_tour", method: :put, class: 'btn btn-info btn-block' %>
```
Comment (talyssonoc) [inline app/views/registrations/edit.html.erb]: You can use the helper to create the path here, avoid hardcoded URLs like this.

File: package.json [hunk 23]
```diff
@@ -7,6 +7,7 @@
     "coveralls": "cat ./js_coverage/lcov.info | ./node_modules/coveralls/bin/coveralls.js && rm -rf ./js_coverage"
   },
   "dependencies": {
+    "axios": "^0.15.3",
```
Comment (talyssonoc) [inline package.json]: We already have jQuery on the project, why not use `$.ajax`? It's compliant with Axios interface.

File: app/controllers/registrations_controller.rb (hunk 1 above)
Comment (marclerodrigues) [inline app/controllers/registrations_controller.rb]: I believe you could just do:
```ruby
class UsersController < ApplicationController
 before_filter :set_project, except: [:tour, :reset_tour]
end
```

File: package.json (hunk 23 above)
Comment (FidelisClayton) [inline package.json]: I was having some problems with ajax, but it's ok now!

File: app/views/layouts/application.html.erb (hunk 18 above)
Comment (FidelisClayton) [inline app/views/layouts/application.html.erb]: It's necessary to set the i18n current locale in javascript.

https://github.com/fnando/i18n-js#setting-up

Comment (talyssonoc) [inline app/views/layouts/application.html.erb]: It should be already be done automatically by this: https://github.com/Codeminer42/cm42-central/blob/master/app/views/layouts/application.html.erb#L114, weird o_O

File: app/controllers/registrations_controller.rb (hunk 1 above)
Comment (FidelisClayton) [inline app/controllers/registrations_controller.rb]: It works! Thanks!!!

Comment (FidelisClayton) [inline app/controllers/registrations_controller.rb]: Actually, it's returning me an error:
```
Pundit::AuthorizationNotPerformedError at /users/16/tour
========================================================

> UsersController

app/controllers/application_controller.rb, line 40
--------------------------------------------------

   35       options = [session[:locale], current_user&.locale, 'en']
   36       I18n.locale = (options & I18n.available_locales.map(&:to_s)).first
   37     end
   38   
   39     def user_time_zone(&block)
>  40       Time.use_zone(current_user.time_zone, &block)
   41     end
   42   
   43     def user_not_authorized
   44       flash[:error] = t('users.You are not authorized to perform this action')
   45       redirect_to request.headers["Referer"] || root_path
```

File: app/assets/javascripts/tour.js (hunk 15 above)
Comment (FidelisClayton) [inline app/assets/javascripts/tour.js]: It's because `tour_steps` is stored as a `text`.

File: app/controllers/registrations_controller.rb (hunk 1 above)
Comment (marclerodrigues) [inline app/controllers/registrations_controller.rb]: You should check this error:
http://stackoverflow.com/questions/35071428/punditauthorizationnotperformederror
You must authorize your resource.

File: app/assets/javascripts/models/user.js [hunk 24]
```diff
@@ -1,3 +1,15 @@
...
+  },
+  getCurrent: function(callback) {
+    var options = {
+      type: 'GET',
+      dataType: 'json',
+      url: '/users/current',
+      success: callback,
+    };
+
+    return $.ajax(options);
```
Comment (talyssonoc) [inline app/assets/javascripts/models/user.js]: What do you think about using a promise here? Then, on the call site, you'll just do `user.getCurrent().then(getCurrentUserSuccess)`, I think it's much cleaner and avoid the creation of a parameter for the callback 😄

File: app/assets/javascripts/models/tour.js [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+  create: function() {
+    return new Shepherd.Tour({
+      defaults: {
+        classes: 'shepherd-theme-arrows',
+        scrollTo: false
+      }
+    })
+  },
+  update: function(user) {
+    var remainingSteps = user.tour_steps.filter(step => step.done !== true);
```
Comment (talyssonoc) [inline app/assets/javascripts/models/tour.js]: We should avoid this lambda here until @eddiecode's PR is merged.

File: app/assets/javascripts/models/user.js [hunk 26]
```diff
@@ -1,3 +1,15 @@
 module.exports = Backbone.Model.extend({
-  name: 'user'
+  name: 'user',
+  initialize: function() {
```
Comment (talyssonoc) [inline app/assets/javascripts/models/user.js]: Is it necessary to add a constructor to this model?

File: app/assets/javascripts/models/user.js (hunk 24 above)
Comment (talyssonoc) [inline app/assets/javascripts/models/user.js]: Isn't it supposed to return a user instance? Why not add a _static_ method that: does the ajax, creates a new user instance, and then return it in a promise? There's no reason to create a user instance and then call a `getCurrent` on it to return another user, it's counterintuitive. We could do something like this:

```js
var User = module.exports = Backbone.Model.extend({
    name: 'user'
  },
  { // << notice that this is the _second_ parameter of `extend`, that receives the static methods
    getCurrent: function(callback) {
      var options = {
        type: 'GET',
        dataType: 'json',
        url: '/users/current'
    };

    return $.ajax(options).then(function(userData) {
      return new User(userData);
    });
  }
});

// Then you'd do something like this:

User.getCurrent().then(getCurrentUserSuccess);
```

File: app/assets/javascripts/models/tour.js [hunk 27]
```diff
@@ -0,0 +1,38 @@
+var Tether = require('tether');
+var Shepherd = require('tether-shepherd');
+
+module.exports = Backbone.Model.extend({
+  name: 'tour',
+  initialize: function(){
+  },
+  create: function() {
+    return new Shepherd.Tour({
```
Comment (talyssonoc) [inline app/assets/javascripts/models/tour.js]: What is the rationale behind having a `create` method in a instance? I think all those methods from `Shepherd.Tour` could be encapsulated on the `Tour` instance. Suggestions:

- Move the creation of the `Shepher.Tour` instance to the `Tour` model constructor (like `this.shepherd = new Shepher.Tour...`;
- Add `start`, `next` and `complete` instance methods to the `Tour` model that delegates to `this.shepherd`.

File: app/assets/javascripts/tour.js [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+      return buttons;
+    }
+
+    function getSelectorFromStepAttach(attach) {
+      return attach.split(" ")[0];
+    }
+
+    function setupTour(user, tour) {
+      user.tour_steps.forEach((step, index) => {
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: We should avoid this lambda here.

File: app/assets/javascripts/tour.js [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+        var tourStep = Object.assign({}, step);
+
+        var selector = getSelectorFromStepAttach(step.attachTo);
+
+        if($(selector).length > 0 && step.done !== true) {
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: I'd change `step.done !== true` to `!step.done`

File: app/assets/javascripts/models/tour.js (hunk 25 above)
Comment (talyssonoc) [inline app/assets/javascripts/models/tour.js]: Also, I'd change `step.done !== true` to `!step.done`

File: app/assets/javascripts/models/user.js (hunk 26 above)
Comment (FidelisClayton) [inline app/assets/javascripts/models/user.js]: If I don't add the constructor it will return me `Uncaught TypeError: Cannot read property 'hasOwnProperty' of undefined` when I try to access the `getCurrent` function.

Comment (talyssonoc) [inline app/assets/javascripts/models/user.js]: I think this would solve the problem https://github.com/Codeminer42/cm42-central/pull/106#discussion_r99101087

Comment (FidelisClayton) [inline app/assets/javascripts/models/user.js]: I was thinking the same, but it keeps giving me the same error if I remove the `initialize`.

File: app/assets/javascripts/tour.js [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+          tourStep.buttons = generateButtons(index, user);
+
+          tour.addStep(tourStep.title.toLowerCase().replace(' ', '_'), tourStep);
+        }
+      });
+    };
+
+    function getCurrentUserSuccess(user) {
+      var currentUser = user.attributes;
+      currentUser.tour_steps = JSON.parse(currentUser.tour_steps);
```
Comment (talyssonoc) [inline app/assets/javascripts/tour.js]: http://backbonejs.org/#Model-parse

**Comment by talyssonoc** [comment]
We should included that on the CHANGELOG.md (if there's not one already, create one, so the ones who host an instance of the app can keep track of what will be changed).

File: app/assets/stylesheets/application.scss [hunk 31]
```diff
@@ -6,6 +6,7 @@
 @import "compass/css3";
 @import "bootstrap-sprockets";
 @import "bootstrap";
+@import "tether-shepherd/dist/css/shepherd-theme-arrows";
```
Comment (daniloisr) [inline app/assets/stylesheets/application.scss]: would be nice to add a comment before this line to mark this import as a `node_modules` import. To make it easy when we start moving css to webpack.
I thought in something like:
```scss
// node_modules imports
@import "tether-shepherd/dist/css/shepherd-theme-arrows";
```

File: app/controllers/registrations_controller.rb (hunk 1 above)
Comment (daniloisr) [inline app/controllers/registrations_controller.rb]: You need to add the actions to this policy:
https://github.com/Codeminer42/cm42-central/blob/master/app/policies/user_policy.rb#L20

I suggest using the check `is_himself?`. Also, as this action updates only the current user, you should use `current_user` instead of `@user = User.find(params[:id])`

File: app/models/user.rb [hunk 32]
```diff
@@ -47,6 +49,10 @@ def set_reset_password_token
     raw
   end
 
+  def set_tour_steps
+    WelcomeTour.new.set_user_steps self
```
Comment (daniloisr) [inline app/models/user.rb]: This is more easy to see that you are initializing a user attributes:
```ruby
def init_tour_steps
  self.tour_steps = WelcomeTour::STEPS.to_json
end
```
and you can remove the method `set_user_steps` from WelcomeTour.

**Comment by andrerpbts** [comment]
@FidelisClayton I need a squash and a rebase here to merge, can you provide it please?

**Comment by andrerpbts** [comment]
@FidelisClayton btw, Is this still a WIP PR?

**Comment by akitaonrails** [comment]
@FidelisClayton how is this PR going? can you remerge/retest so we can merge?

File: app/controllers/sessions_controller.rb [hunk 33]
```diff
@@ -0,0 +1,9 @@
+class SessionsController < Devise::SessionsController
+  def current
+    if user_signed_in?
+      render json: current_user, status: :ok
+    else
+      render json: {}, status: :unauthorized
```
Comment (adbatista) [inline app/controllers/sessions_controller.rb]: `head :unauthorized`

**Comment by marclerodrigues** [comment]
Please add specs to Registration and Session controllers.

**Comment by andrerpbts** [comment]
Moving it to an own branch of this project, and opening a new PR: https://github.com/Codeminer42/cm42-central/pull/156.

Closing.
